library/bn_mul.h

```diff
@@ -727,11 +743,12 @@
          : "=m" (c),  "=m" (d), "=m" (s)        \
          : "m" (s), "m" (d), "m" (c), "m" (b)   \
          : "r0", "r1", "r2", "r3", "r4", "r5",  \
-           "r6", "r7", "r8", "r9", "cc"         \
+           "r6", MULADDC_SCRATCH_CLOBBER, "r8", "r9", "cc" \
          );
 
 #elif (__ARM_ARCH >= 6) && \
     defined (__ARM_FEATURE_DSP) && (__ARM_FEATURE_DSP == 1)
+/* Armv6-M with DSP Instruction Set Extensions */
```

Comment:
Armv6 or greater. This branch also requires Thumb-2 or Arm instructions for the post-indexed load -- it might be better to move the `&& !defined(__ARMCC_VERSION)` part of the first condition to a nested #if to prevent any Thumb-1-only CPUs dropping into the else cases at all.

Reply:
I've done this, but I think it might not be quite right. We currently use `__thumb__` to detect whether we want to use the Thumb1 codepath, but actually this means "the compiler is targeting Thumb", not "the CPU supports Thumb" which is maybe what we really want? Similarly `__thumb2__`.

Arm docs indicate that `__ARM_ARCH_ISA_ARM` and `__ARM_ARCH_ISA_THUMB` work for this (https://developer.arm.com/documentation/101028/0012/5--Feature-test-macros ) - these are also set by gcc, but not by clang.

This leaves a gap if the user is targeting Thumb on a Thumb-1-and-Arm core, using armclang. Here they could use the third code-path (generic arm) but would end up getting no optimised assembly.

AFAICT, the only such cores are very old (all M-class cores don't support the Arm ISA). So I'm not going to create more churn and complexity in the ifdefs to try and cover this.

Reply:
We do want to know “the compiler is targeting Thumb”. If the compiler is emitting Arm instructions, it's irrelevant whether the CPU supports Thumb instructions. So we don't care about `__ARM_ARCH_ISA_THUMB`.

For all of GCC, clang and armclang (at least the recent-ish versions I tested), the differences between compiling for thumb vs arm with `-mcpu=cortex-a8` are that in thumb mode, the following extra macros are defined:
```
> #define __thumb2__ 1
> #define __thumb__ 1
> #define __THUMBEL__ 1
```
(plus `#define __ARM_ASM_SYNTAX_UNIFIED__ 1` with GCC, not relevant here I think.)